### PR TITLE
[MRG] fix bug in `LCA_Database.downsample`

### PR DESCRIPTION
**Pull request by ctb**
I discovered this bug while working on other things -

In the base `LCA_Database` implementation, a dictionary `_hashval_to_idx` is used to track which hash values (keys) belong to which set of signatures (individual signatures are referenced by integers, `idx`). The dictionary is created as a `collections.defaultdict(set)` [ref](https://docs.python.org/3/library/collections.html#collections.defaultdict) so that when adding a new key, it automatically comes with a new set as a value - this permits the call
```python
self._hashval_to_idx[hashval].add(idx)
```
in `LCA_Database.insert(...)`.

The bug is that after `LCA_Database.downsample(...)` is called, `_hashval_to_idx` is recreated as a regular dictionary, which causes `insert` to fail.

This PR adds a new test, `test_lca.py:test_api_create_insert_two_then_scale_then_add`, that does a downsample and _then_ does an insert. This test exposes the bug above, which the PR also fixes.

### CODE REVIEW

**Comment by ctb** [comment]
@ccbaumler can you take on reviewing this? Should be enough to verify that the description of the PR matches the code changes, and that the tests pass. No hurry. Thanks!

**Comment by ccbaumler** [comment]
I'm on it!

**Comment by ccbaumler** [comment]
@ctb Two questions:

1. Why use `new_hashvals = defaultdict(set)` instead of `new_hashvals = defaultdict(list)`? Would the `list` data structure provide the indices for the hash values? 
2. In the function `test_api_create_insert_two_then_scale_then_add()`, should the other `ksize` variables and changes in `scale` be accounted for?

**Comment by ctb** [comment]
> @ctb Two questions:
> 
>     1. Why use `new_hashvals = defaultdict(set)` instead of `new_hashvals = defaultdict(list)`? Would the `list` data structure provide the indices for the hash values?

`set` is a good data structure to use for situations where you have a container of element and:

* you don't care about ordering of elements;
* you are only asking about presence/absence, not multiplicity (number of times something is seen);

and they are optimized for membership checks.

`list`, by contrast,

* maintain order of elements
* will store multiple identical elements

and they are not optimized for a membership check, but rather for iteration (`for item in listx:`) and indexing (`item = listx[i]`).

>     2. In the function `test_api_create_insert_two_then_scale_then_add()`, should the other `ksize` variables and changes in `scale` be accounted for?

this is hard to respond to comprehensively, because it enters a whole world of testing philosophy, but here are a few thoughts -

(1) a good rule for a FIRST test in a bugfix PR is one that replicates the bug precisely.
(2) I strongly believe that there is no interaction in this particular aspect of the code with ksize and scaled.
(3) generally speaking I usually want a specific reason to add a test - to increase code statement coverage, or branch coverage;

that all having been said, if you have reason to believe that I'm wrong - especially about point (2) - the ideal path forward would be to describe why you think we need additional tests, or -- even better!  -- to write a test that breaks something that should work :).

tl;dr we can always write more tests, but it's good to have a specific reason for doing so, however nebulous.

**Comment by ctb** [comment]
(great questions!)

**Review by ccbaumler — APPROVED** [review]
Thanks for the insight! LGTM.
